### PR TITLE
Crop to physical space

### PR DESCRIPTION
**Pull request by joeranbosma**
Allow crop/pad of scans based on physical size

### CODE REVIEW

**Comment by snorthman** [comment]
Before review, can you explain the failing tests?
`tests/test_mha2nnunet.py FF...                                           [ 90%]`

**Comment by joeranbosma** [comment]
Those failed tests are from the second-to-last commit I think. There I made a mistake in the zyx/xyz order convention, which I fixed later. The PR has all tests passing.

File: src/picai_prep/preprocessing.py [hunk 1]
```diff
@@ -118,16 +118,35 @@ def resample_img(
...
+
+    Either size or physical_size must be provided.
+
+    Returns:
+    - resized image (same type as input)
     """
     # input conversion and verification
+    if physical_size is not None:
+        # convert physical size to voxel size (only supported for SimpleITK)
+        assert isinstance(image, sitk.Image), "Crop/padding by physical size is only supported for SimpleITK images."
```
Comment (snorthman) [inline src/picai_prep/preprocessing.py:140]: assert should not be used in production code. use if/raise instead.

**Review by snorthman — CHANGES_REQUESTED** [review]

File: src/picai_prep/preprocessing.py (hunk 1 above)
Comment (joeranbosma) [reply src/picai_prep/preprocessing.py]: I agree. Made a new commit to tackle this. Also, added a new function `crop` where a maximum size can be specified. I need this functionality, do you think this is the right place to store this function? (as it is not used inside the picai_prep module itself, but does pertain to preprocessing).

File: src/picai_prep/preprocessing.py [hunk 2]
```diff
@@ -357,7 +387,11 @@ def preprocess(self):
             self.resample_spacing()
 
         if self.settings.matrix_size is not None or self.settings.physical_size is not None:
-            # perform centre crop
+            # perform centre crop and/or pad
+            self.centre_crop_or_pad()
+
+        if self.settings.max_matrix_size is not None or self.settings.max_physical_size is not None:
```
Comment (snorthman) [inline src/picai_prep/preprocessing.py:394]: while I know in the post_init neither matrix or max_matrix can be both set, I still think its safer to change this to `elif`

**Review by snorthman — COMMENTED** [review]
this does have a smell of feature creep. suggest to instead create a setting
`crop_only`, making physical_size and/or matrix_size act as the current max_matrix_size, ... instead

**Comment by joeranbosma** [comment]
Thanks for checking out the changes! I agree fully, implementing the `crop_only` variant now.

File: src/picai_prep/preprocessing.py [hunk 3]
```diff
@@ -421,29 +362,39 @@ def preprocess(self):
...
+            raise ValueError("Need SimpleITK Image or reference scan for phsyical metadata!")
         image: sitk.Image = sitk.GetImageFromArray(image)
-        assert reference_scan_preprocessed is not None, "Need reference scan for phsyical metadata!"
 
     if reference_scan_preprocessed is not None:
         image.CopyInformation(reference_scan_preprocessed)
 
+    if interpolator is None:
+        # determine interpolation method based on image dtype
+        if "integer" in image.GetPixelIDTypeAsString():
```
[the remark this answers is not in the document]
Comment (joeranbosma) [reply src/picai_prep/preprocessing.py]: Done

**Pull request merged**